cloud/scope/powervs_cluster.go

```diff
@@ -882,37 +915,80 @@ func (s *PowerVSClusterScope) ReconcileNetwork() (bool, error) {
 	return false, nil
 }
 
-// checkNetwork checks if network exists in cloud with given name or ID mentioned in spec.
-func (s *PowerVSClusterScope) checkNetwork() (*string, error) {
-	if s.IBMPowerVSCluster.Spec.Network.ID != nil {
-		s.V(3).Info("Checking if PowerVS network exists in IBM Cloud with ID", "networkID", *s.IBMPowerVSCluster.Spec.Network.ID)
-		network, err := s.IBMPowerVSClient.GetNetworkByID(*s.IBMPowerVSCluster.Spec.Network.ID)
+// checkDHCPServer checks if DHCP server exists in cloud with given DHCPServer's ID or name mentioned in spec.
+// If exists and s.IBMPowerVSCluster.Status.Network is not populated will set DHCP server's network as cluster's network.
+// If exists and s.IBMPowerVSCluster.Status.Network is populated already will validate the DHCP server's network and cluster networks are matching, if not will throw an error.
+func (s *PowerVSClusterScope) checkDHCPServer() (*string, error) {
+	if s.DHCPServer() != nil && s.DHCPServer().ID != nil {
+		dhcpServer, err := s.IBMPowerVSClient.GetDHCPServer(*s.DHCPServer().ID)
 		if err != nil {
 			return nil, err
 		}
-		return network.NetworkID, nil
+		if s.GetNetworkID() == nil {
+			if _, err := s.IBMPowerVSClient.GetNetworkByID(*dhcpServer.Network.ID); err != nil {
+				return nil, err
+			}
+			s.SetStatus(infrav1beta2.ResourceTypeNetwork, infrav1beta2.ResourceReference{ID: dhcpServer.Network.ID, ControllerCreated: ptr.To(false)})
+		} else if *dhcpServer.Network.ID != *s.GetNetworkID() {
```

Comment:
can there occur a scenario where `dhcpServer.Network` is nil?

Reply:
I don't think so, but have added a nil check for `dhcpServer.Network` anyway, ptal.